reference/best-practices/java-app.md

```diff
@@ -0,0 +1,362 @@
+
+```xml
+<insert id="insertTestBatch" parameterType="java.util.List" fetchSize="1">
+  insert into test
+   (id, v1, v2)
+  values
+  <foreach item="item" index="index" collection="list" separator=",">
+  (
+   #{item.id}, #{item.v1}, #{item.v2}
+  )
+  </foreach>
+  on duplicate key update v2 = v1 + values(v1)
+</insert>
+```
+
+This mapper generates an `insert on duplicate key update` statement. The number of `(?,?,?)` following "values" is determined by the number of passed lists. Its final effect is similar to using `rewriteBatchStatements=true`, which also effectively reduces communication overhead between the client and TiDB.
+
+As mentioned before, you also need to note that the Prepared Statements will not be cached after their maximum length exceeds the value of `prepStmtCacheSqlLimit`.
+
+#### Streaming result
+
+The earlier section introduces how to stream read execution results in JDBC. In addition to the corresponding configurations of JDBC, if you want to read a super large result set in MyBatis, you also need to note that:
+
+- You can set `fetchSize` for a single SQL statement in the mapper configuration (see the previous code block). Its effect is equivalent to calling `setFetchSize` in JDBC.
+- You can use the query interface with `ResultHandler` to avoid getting the entire result set at once.
+- You can use the `Cursor` class for stream reading.
+
+If you configure mappings using XML, you can stream read results by configuring `fetchSize="-2147483648"`(`Integer.MIN_VALUE`) in the mapping's `<select>` section.
+
+```xml
+<select id="getAll" resultMap="postResultMap" fetchSize="-2147483648">
+  select * from post;
+</select>
+```
+
+If you configure mappings using code, you can use `@Options(fetchSize = Integer.MIN_VALUE)` and return `Cursor` so that the SQL results can be read in streaming.
+
+```java
+@Select("select * from post")
+@Options(fetchSize = Integer.MIN_VALUE)
+Cursor<Post> queryAllPost();
+```
+
+### `ExecutorType`
+
+You can choose `ExecutorType` during `openSession`. MyBatis supports three types of executors:
+
+- Simple: The Prepared Statements are called to JDBC for each execution (if the JDBC configuration item `cachePrepStmts` is enabled, repeated Prepared Statements will be reused)
+- Reuse: The Prepared Statements are cached in `executor`, so that you can reduce duplicate calls for Prepared Statements without using the JDBC `cachePrepStmts`
+- Batch: Each update statement will only be executed with `ExecuteBatch()` called when `addBatch()` is used for a query or a commit. If `rewriteBatchStatements` is enabled in the JDBC layer, it will try to rewrite the statements. If not, the statements will be sent one by one.
```

Comment:
```suggestion
- Batch: Each update operation (`INSERT`/`DELETE`/`UPDATE`) will first be added to the batch, and will be executed until the transaction commits or a `SELECT` query is performed. If `rewriteBatchStatements` is enabled in the JDBC layer, it will try to rewrite the statements. If not, the statements will be sent one by one.
```